Add a Robot class for Move to Pose Algorithm

Muhammad-Yazdian opened the pull request:
#### What does this implement/fix?
- An improvement on move_to_pose.py
- I implemented a Robot class to organize the move to pose algorithm (see image below)
![Figure_1](https://user-images.githubusercontent.com/93126501/145834505-a8df8311-5445-413f-a96f-00460d47991c.png)

#### Additional information
The advantage of this class is that you can define as many as robots with different specifications (name, color, controller and speed limitations)

#### CheckList
- the unittest was passed.
- The documents text added. Where should I upload the gif/image?

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,242 @@
+
+import matplotlib.pyplot as plt
+import numpy as np
+from random import random
```
‘random’ is not used?

Muhammad-Yazdian replied:
No. I just removed it.

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,242 @@
+
+        if abs(angular_velocity) > self.MAX_ANGULAR_SPEED:
+            angular_velocity = (np.sign(angular_velocity)
+                                * self.MAX_ANGULAR_SPEED)
```
I think this function has duplicated codes in ‘ move_to_pose.py’. I think some common function should be created in ‘ move_to_pose.py’, and this file should use these.

Muhammad-Yazdian replied:
No. the move_to_pose_robot_class.py and move_to_pose.py have totally different structures. I rewrote move_to_pose.py and turned it to move_to_pose_robot_class.py to make its functionalities more modular. I can create separate class module so that everyone can simply import it in his/her main script. Please let me know if you would like me to do that?

AtsushiSakai replied:
I think you can move the `PathFinderController` to the `move_to_pose.py` and it has control logic. It is better taht this script imports it and uses it.

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,242 @@
+    global all_robots_are_at_target
+    global simulation_running
+
+    PLOT_WINDOW_SIZE_X = 20
```
These const variables should go top of this file?

Muhammad-Yazdian replied:
- We need 'global all_robots_are_at_target' and 'global simulation_running' here. 
  - Note that these two global variables were already defined at line ~23 and ~24.
  - They are runtime flags. Aka. we need to _update_ them locally during simulation in order to terminate the simulation.
- I moved PLOT_WINDOW_SIZE_X = 20 to top of the file.

AtsushiSakai commented on docs/modules/control/move_to_a_pose_control/move_to_a_pose_control.rst:
```diff
@@ -5,6 +5,12 @@ This is a simulation of moving to a pose control
+implemented Robot class where you can define your robots and simulate robots 
+with differnt specificaitons at the same time.
+
+.. image:: https://user-images.githubusercontent.com/93126501/145834505-a8df8311-5445-413f-a96f-00460d47991c.png
```
If you want add a gif, please submit a PR to https://github.com/AtsushiSakai/PythonRoboticsGifs/

Muhammad-Yazdian replied:
Ok. I will try to upload one after finishing this PR.

AtsushiSakai commented on docs/modules/control/move_to_a_pose_control/move_to_a_pose_control.rst:
```diff
@@ -5,6 +5,12 @@ This is a simulation of moving to a pose control
 
 .. image:: https://github.com/AtsushiSakai/PythonRoboticsGifs/raw/master/PathTracking/move_to_pose/animation.gif
 
+For a more advanced version, see move_to_pose_robot_class.py. The program 
```
If possible, could you please add a mathematical note of this controller?

Muhammad-Yazdian replied:
Yes. I was working on it yet. One thing that I notice is that the Kp_beta must be negative which does not make sense.

AtsushiSakai replied:
If you want, you can fix #600 in this PR.

Muhammad-Yazdian replied:
I fixed reviewed the issue #600 and committed it in this PR. It was kinda difficult to catch it. I explain the logic here: Both 'Kp_alpha * alpha' and  'Kp_beta * beta' are significant values even at the vicinity of the target. When rho = ~0 and the heading is not aligned, the combined terms (Kp_alpha * alpha - Kp_beta * beta') rotates the robot toward the target direction. The reason that we need a negative sign for Kp_beta is that when we are far from the target we need an offset in the opposite direction so can help the robot to smoothly approach the target direction otherwise if Kp_beta = 0 the robot directly goes to the target (x,y) position. Another way to think about the negative sign of the Kp_beta is the way we take turns bicycle. For example, we initially lean left when we want to take a right turn. Indeed for the bicycle turning, the main reason is due to the dynamic motion. It was just a simple analogy to explain -Kp_beta easier...

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,242 @@
+    p3_i = np.array([-0.5, -0.25, 1]).T
+
+    T = transformation_matrix(x, y, theta)
+    p1 = np.matmul(T, p1_i)
```
Could please use @ instead of matmul?

Muhammad-Yazdian replied:
The matmul methods changed to @.

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,242 @@
+Move to specified pose (with Robot class)
+
+Author: Daniel Ingram (daniel-s-ingram)
+        Atsushi Sakai(@Atsushi_twi)
```
Could you add a space after sakai?

Muhammad-Yazdian replied:
Done. :)

Muhammad-Yazdian commented:
- The Linux_CI says 
  - The test for 'test_probabilistic_road_map.py::test1' failed. 
  - The 'move_to_pose_robot_class.py' test passed.
  - Do we need to run all test for this PR?

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,240 @@
+# Simulation parameters
+TIME_DURATION = 1000
+TIME_STEP = 0.01
+AT_TARGET_ACCEPTANCE_THRESHOULD = 0.01
```
THRESHOLD?

Muhammad-Yazdian replied:
Gotcha. Thanks!

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,240 @@
+        y_diff = self.pose_target.y - self.pose.y
+        rho = np.hypot(x_diff, y_diff)
+
+        if(rho < AT_TARGET_ACCEPTANCE_THRESHOULD):
```
redundant parentheses

Muhammad-Yazdian replied:
Fixed.

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,240 @@
+        robot_names.append(instance.name)
+
+    time = 0
+    while (simulation_running and time < TIME_DURATION):
```
redundant parentheses

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,240 @@
+        robots_are_at_target = []
+
+        for instance in robots:
+            if(not instance.is_at_target):
```
redundant parentheses

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,240 @@
+    robot_2.set_start_target_poses(pose_start_2, pose_target)
+    robot_3.set_start_target_poses(pose_start_3, pose_target)
+
+    robots: list[Robot] = []
+    robots.append(robot_1)
+    robots.append(robot_2)
+    robots.append(robot_3)
```
```suggestion
    robots: list[Robot] = [robot_1, robot_2, robot_3]
```

AtsushiSakai commented on Control/move_to_pose/move_to_pose_robot_class.py:
```diff
@@ -0,0 +1,229 @@
+        if rho < AT_TARGET_ACCEPTANCE_THRESHOLD:
+            self.is_at_target = True
+
+        # Restrict alpha and beta (angle differences) to the range
+        # [-pi, pi] to prevent unstable behavior e.g. difference going
+        # from 0 rad to 2*pi rad with slight turn
+
+        alpha = (np.arctan2(y_diff, x_diff)
+                 - self.pose.theta + np.pi) % (2 * np.pi) - np.pi
+        beta = (self.pose_target.theta - self.pose.theta - alpha
+                + np.pi) % (2 * np.pi) - np.pi
+
+        linear_velocity = self.path_finder_controller.Kp_rho * rho
+        angular_velocity = (self.path_finder_controller.Kp_alpha * alpha
+                            + self.path_finder_controller.Kp_beta * beta)
+
+        if alpha > np.pi / 2 or alpha < -np.pi / 2:
+            linear_velocity = -linear_velocity
+
+        if abs(linear_velocity) > self.MAX_LINEAR_SPEED:
+            linear_velocity = (np.sign(linear_velocity)
+                               * self.MAX_LINEAR_SPEED)
+
+        if abs(angular_velocity) > self.MAX_ANGULAR_SPEED:
+            angular_velocity = (np.sign(angular_velocity)
+                                * self.MAX_ANGULAR_SPEED)
```
I think these control logics should be in the `PathFinderController`.

Muhammad-Yazdian replied:
I added update_command() method to the PathFinderController class. Note that the velocity check is not part of the path finder task so I did not move them to the update_command(). The pathFinderController just tells the driver how to steer the wheel and how to push on gas. It does not care about the limitation of the car/robot.

AtsushiSakai commented on Control/move_to_pose/move_to_pose.py:
```diff
@@ -13,10 +13,38 @@
+        self.Kp_alpha = Kp_alpha
+        self.Kp_beta = Kp_beta
+
+    def update_command(self, x_diff, y_diff, theta, theta_goal):
```
`update_command` -> `calc_control_command` is better?

AtsushiSakai commented on Control/move_to_pose/move_to_pose.py:
```diff
@@ -13,10 +13,38 @@
+        self.Kp_beta = Kp_beta
+
+    def update_command(self, x_diff, y_diff, theta, theta_goal):
+        """Updates the command linera and angular velocities"""
```
Could you add documentation to explain about input and output of this function?

AtsushiSakai reviewed (changes requested):
Thank you. This is the last review. After addressing these my comments, I will merge it.

Muhammad-Yazdian commented:
@AtsushiSakai I think the whole move to pose code algorithm is in a great shape right now. I rename one file; added docstring comments; and updated the documentation. To expedite the process, I have two questions:

1. What tool would you suggest to use for removing trailing whitespaces? I'm using VS Code and it seems that the built-in 'Format Document' feature does not remove trailing whitespaces inside comment/docstring texts.
2. Is there any way that I can preview my the doc after I modify some of its contents?

Thank you

AtsushiSakai commented:
> What tool would you suggest to use for removing trailing whitespaces? I'm using VS Code and it seems that the built-in 'Format Document' feature does not remove trailing whitespaces inside comment/docstring texts.

I'm not using any tool to remove trailing white spaces in comment/docstring texts. I will try to find a format tool for the issue.

> Is there any way that I can preview my the doc after I modify some of its contents?

Usually, you can check the latest doc in Circle CI result [Ref](https://pythonrobotics.readthedocs.io/en/latest/how_to_contribute.html#step-5-submit-a-pull-request-and-fix-codes-based-on-review)
But, It seems that the latest CI result does have it, maybe if you commit something, it will be shown.

AtsushiSakai commented on docs/modules/control/move_to_a_pose_control/move_to_a_pose_control.rst:
```diff
@@ -1,11 +1,102 @@
 
-Ref:
 
+PathFinderController class
```
[General comment] I think the rst doc should focus on the mathematics and the algorithm of the example.  Documentation of codes should be in python script as docstrings or header comments. If you have time, please consider to fix the doc in the next PR.

Muhammad-Yazdian replied:
@AtsushiSakai  Thanks for your comments. I agree to keep the rst doc only for the mathematical explanations. Also, it will be great if we can automatically generate a Reference document based on the docstrings comments. I will try to address  your points in my next PR. 

Merry Christmas and Happy new Year :)

AtsushiSakai reviewed (approved):
I think the doc needs improvement, but the next PR is fine. Let's merge this PR. Thank you for your work!!
